rules/framework/vfs_overlay.bzl

```diff
@@ -1,40 +1,196 @@
+        }]
+
+    private_headers_contents = {
+        "name": "PrivateHeaders",
         "type": "directory",
         "contents": [
             {
                 "type": "file",
-                "name": override_name or file.basename,
+                "name": file.basename,
                 "external-contents": file.path,
             }
-            for file in files
+            for file in private_hdrs
         ],
     }
 
+    private_headers = []
+    if len(private_hdrs):
+        private_headers = [private_headers_contents]
+
+    roots = []
+    if len(headers) or len(private_headers) or len(modules):
+        roots += [{
+            "name": root_dir,
+            "type": "directory",
+            "contents": headers + private_headers + modules,
+        }]
+    if has_swift:
+        roots += [_vfs_swift_module_contents(ctx, framework_name, FRAMEWORK_SEARCH_PATH)]
+
+    return roots
+
+def _vfs_swift_module_contents(ctx, framework_name, root_dir):
+    # Forumlate the framework's swiftmodule - don't have the swiftmodule when
+    # creating with apple_library. Consider removing that codepath to make this
+    # and other situations easier
+    base_path = "/".join(ctx.build_file_path.split("/")[:-1])
+    bin_dir = ctx.bin_dir.path
+    rooted_path = bin_dir + "/" + base_path
+
+    # Note: Swift translates the input framework name to this because - is an
+    # invalid module name
+    name = framework_name.replace("-", "_") + ".swiftmodule"
+    external_contents = rooted_path + "/" + name
+    return {
+        "type": "file",
+        "name": root_dir + "/" + name,
+        "external-contents": external_contents,
+    }
+
 def _framework_vfs_overlay_impl(ctx):
+    vfsoverlays = []
+
+    # Conditionally collect and pass in the VFS overlay here.
+    virtualize_frameworks = feature_names.virtualize_frameworks in ctx.features
+    if virtualize_frameworks:
+        for dep in ctx.attr.deps:
+            if FrameworkInfo in dep:
+                vfsoverlays.extend(dep[FrameworkInfo].vfsoverlay_infos)
+            if VFSOverlayInfo in dep:
+                vfsoverlays.append(dep[VFSOverlayInfo].vfs_info)
+
+    vfs_ret = make_vfsoverlay(
+        ctx,
+        hdrs = ctx.files.hdrs,
+        module_map = ctx.files.modulemap,
+        private_hdrs = ctx.files.private_hdrs,
+        has_swift = ctx.attr.has_swift,
+        merge_vfsoverlays = vfsoverlays,
+        output = ctx.outputs.vfsoverlay_file,
+        extra_search_paths = ctx.attr.extra_search_paths,
+    )
+
+    headers = depset([vfs_ret.vfsoverlay_file])
+    cc_info = CcInfo(
+        compilation_context = cc_common.create_compilation_context(
+            headers = headers,
+        ),
+    )
+    return [
+        apple_common.new_objc_provider(),
+        cc_info,
+        VFSOverlayInfo(
+            files = depset([vfs_ret.vfsoverlay_file]),
+            vfs_info = vfs_ret.vfs_info,
+        ),
+
+        # Mainly a debugging mechanism
```

Comment:
u mean this is optional and it contains the list of generated json files correct?

Reply:
Yeah maybe we can just remove this now, in an early rev, it had many vfs, so now there's only 1